### PR TITLE
CANS-301: CANS API endpoints secured

### PR DESCRIPTION
**Pull request by LeonidMarushevskyi**
## JIRA
- https://osi-cwds.atlassian.net/browse/CANS-301

## Description

CANS API endpoints secured

## Tests
- [x] I used TDD to develop the feature
- [x] I have included unit tests
- [x] I have included new acceptance tests or modified existing ones
- [ ] I have included other tests

## Types of changes

- [x] New feature (non-breaking change which adds functionality)
- [ ] Refactoring (No behavioral changes)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] My code follows the code style of this project.
- [x] I ran all my tests locally which were green.
- [x] My code adds no new issues to Code Climate
- [x] I ran all the linters and static analyzers for the project (SonarQube, eslint, rubocop, etc).
- [x] My code is in a stable state ready to be deployable, but not necessarily complete.
- [x] I promise on my honor as a CWDS developer to ensure I don't break the build, to check linters, use best practices, to leave the code in better shape than I found it, and to have a good day.

### CODE REVIEW

File: src/test/java/gov/ca/cwds/cans/rest/resource/AuthorizationResourceTest.java [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+
+import java.io.IOException;
+import javax.ws.rs.client.Entity;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+import org.junit.Test;
+
+public class AuthorizationResourceTest extends AbstractFunctionalTest {
+
+  private static final String START_ASSESSMENT_REQUEST = "gov.ca.cwds.cans.domain.dto.assessment.StartAssessmentRequest";
```
Comment (denys-davydov) [inline src/test/java/gov/ca/cwds/cans/rest/resource/AuthorizationResourceTest.java]: `StartAssessmentRequest.class.getCanonicalName()` can be used instead of the harcoded value here. The same with next 2 lines.

File: src/test/java/gov/ca/cwds/cans/rest/resource/AuthorizationResourceTest.java [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+        PERSON_DTO, HttpMethod.POST, 403);
+
+    callEndpoint(PEOPLE + SLASH + SEARCH, FIXTURES_PERSON_POST_JSON,
+        PERSON_DTO, HttpMethod.POST, 403);
+
+    callEndpoint(PEOPLE + SLASH + ID, null,
+        null, HttpMethod.DELETE, 403);
+  }
+
+  private void callEndpoint(String resourceUrl, String requestFixture, String requestClass,
```
Comment (denys-davydov) [inline src/test/java/gov/ca/cwds/cans/rest/resource/AuthorizationResourceTest.java]: What is your opinion if we call this method `assertEndpointIsSecured` or smth like that, and remove the `expectedResponseCode` parameter since it is always 403?

File: src/test/java/gov/ca/cwds/cans/rest/resource/AuthorizationResourceTest.java (hunk 1 above)
Comment (LeonidMarushevskyi) [reply src/test/java/gov/ca/cwds/cans/rest/resource/AuthorizationResourceTest.java]: Agree. Applied this change.

File: src/test/java/gov/ca/cwds/cans/rest/resource/AuthorizationResourceTest.java (hunk 2 above)
Comment (LeonidMarushevskyi) [reply src/test/java/gov/ca/cwds/cans/rest/resource/AuthorizationResourceTest.java]: Valid point. Applied this change

File: src/test/java/gov/ca/cwds/cans/rest/resource/AuthorizationResourceTest.java [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+
+    callEndpoint(ASSESSMENTS + SLASH + ID, FIXTURE_START,
+        START_ASSESSMENT_REQUEST, HttpMethod.PUT, 403);
+
+    callEndpoint(ASSESSMENTS, FIXTURE_START,
+        START_ASSESSMENT_REQUEST, HttpMethod.POST, 403);
+  }
+
+  @Test
+  public void countiesEndpoint_failed_whenUnauthorizedUser()
```
Comment (edgibbs) [inline src/test/java/gov/ca/cwds/cans/rest/resource/AuthorizationResourceTest.java]: Just for my own edification, do we have a convention we're following for JUnit test names?

File: src/test/resources/fixtures/perry-account/000-all-authorized.json [hunk 4]
```diff
@@ -2,14 +2,15 @@
...
   "county_name": "Marin",
   "county_cws_code": "1088",
   "privileges": [
-    "CWS CANS System",
     "Sensitive Persons",
-    "Sealed"
+    "Sealed",
+    "CANS-rollout"
   ]
 }
```
Comment (edgibbs) [inline src/test/resources/fixtures/perry-account/000-all-authorized.json]: Could cleanup the newline issue.

File: src/main/java/gov/ca/cwds/cans/rest/resource/AssessmentResource.java [hunk 5]
```diff
@@ -69,6 +72,8 @@ public AssessmentResource(
     }
   )
   @ApiOperation(value = "Post new Assessment", response = AssessmentDto.class)
+  @RequiresRoles(CANS_WORKER_ROLE)
```
Comment (edgibbs) [inline src/main/java/gov/ca/cwds/cans/rest/resource/AssessmentResource.java]: These same two annotations end up getting added on every endpoint.  It is very explicit and easy to reason about.  But if we don't require more customized fine-grained endpoint protection, I'm wondering if there's a way to DRY this up? Can this be accomplished in a filter, but that may be Elixir/Ruby or Scala ideas I'm carrying over that don't apply.

File: src/main/java/gov/ca/cwds/cans/rest/auth/CansStaticAuthorizer.java [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+  public static final String CANS_ROLLOUT_PERMISSION = "CANS-rollout";
+  public static final String CANS_WORKER_ROLE = "CANS-worker";
+
+  /**
+   * Implementation of StaticAuthorizer.
+   * @param perryAccount Perry Account
+   * @param simpleAuthInfo Simple Authorization information
+   */
+  @Override
+  public void authorize(PerryAccount perryAccount, SimpleAuthorizationInfo simpleAuthInfo) {
```
[the remark this answers is not in the document]
Comment (LeonidMarushevskyi) [reply src/main/java/gov/ca/cwds/cans/rest/auth/CansStaticAuthorizer.java]: Added unit test for corner cases.

File: src/main/java/gov/ca/cwds/cans/rest/resource/AssessmentResource.java (hunk 5 above)
Comment (LeonidMarushevskyi) [reply src/main/java/gov/ca/cwds/cans/rest/resource/AssessmentResource.java]: Usage of annotations for authorization purposes is a common practice using Shiro security library. We have a particular case where all resources have the same authorization rules. In the general case, different resources can have different permissions.

**Comment by LeonidMarushevskyi** [comment]
@DmitriyRudenko Good catch. Removed code which propagated roles from static authorizer. It works fine.

**Comment by LeonidMarushevskyi** [comment]
@edgibbs @DmitriyRudenko Requested changes are made. Please revisit.

File: src/main/java/gov/ca/cwds/cans/rest/auth/CansStaticAuthorizer.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (edgibbs) [reply src/main/java/gov/ca/cwds/cans/rest/auth/CansStaticAuthorizer.java]: 👍  for the unit tests

File: src/test/java/gov/ca/cwds/cans/rest/auth/CansStaticAuthorizerTest.java [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+
+public class CansStaticAuthorizerTest {
+
+  public CansStaticAuthorizerTest() {
+  }
+
+  private StaticAuthorizer cansStaticAuthorizer = new CansStaticAuthorizer();
+
+  @Test
+  public void testAuthorize_whenUserAuthorized() throws IOException {
```
[the remark this answers is not in the document]
Comment (LeonidMarushevskyi) [reply src/test/java/gov/ca/cwds/cans/rest/auth/CansStaticAuthorizerTest.java]: Removed dependency on fixtures.

File: src/main/java/gov/ca/cwds/cans/rest/resource/AssessmentResource.java [hunk 8]
```diff
@@ -37,8 +39,9 @@
 import javax.ws.rs.Produces;
 import javax.ws.rs.core.MediaType;
 import javax.ws.rs.core.Response;
+import org.apache.shiro.authz.annotation.RequiresPermissions;
+import org.apache.shiro.authz.annotation.RequiresRoles;
 
-/** @author denys.davydov */
```
[the remark this answers is not in the document]
Comment (LeonidMarushevskyi) [reply src/main/java/gov/ca/cwds/cans/rest/resource/AssessmentResource.java]: @author JavaDoc is used across the code base now. We as the team can talk about usage of this annotation and if we don't need it, we can remove it.

**Comment by LeonidMarushevskyi** [comment]
@edgibbs @DmitriyRudenko please revisit the PR and remove your blocks so I can move the story forward.

File: src/main/java/gov/ca/cwds/cans/rest/resource/AssessmentResource.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (edgibbs) [reply src/main/java/gov/ca/cwds/cans/rest/resource/AssessmentResource.java]: cool
